JuliaLang/julia
# [RFC/WIP] NVPTX/Cuda backend support in Julia

## Goal

The long term goal is to enable support for compiling Julia to different accelerator backends like NVPTX and SPIR and in the future SPIR-V and HSAIL. This PR is a rebase of the work done by @maleadt at [maleadt/master](https://github.com/maleadt/Julia) onto the current Julia master. 
It enables the CUDA backend and in combination with [CUDA.jl](https://github.com/vchuravy/CUDA.jl/tree/vc/fixes) it can compile simple CUDA kernels written in Julia as described in the [blog of maleadt](http://blog.maleadt.net/2015/01/15/julia-cuda)

This is still in the early stage and I would like to invite comments and recommendations for how to proceed further. 
## Components
### Target

edf0693 enables the CUDA target, target selection with `@target` and code compilation. This is the center point of this PR. I would like to discuss the approach taken and how we could scale that to several different backends.
### llvmcall

10434e6 enhances the capabilities of llvmcall as described in #8740. This is necessary to be able to call the NVPTX [intrinsics](https://groups.google.com/forum/#!topic/julia-dev/doQUqQabC8A). 
### bitcasts

Julia ignores address space and to make backends work that use different address space our bitcast needs to respect the address space. See  #9423 for more information on the matter.
## TODO:
  - Should the test compare IR output or can we set up a test architecture with [gpuocelot](https://github.com/maleadt/gpuocelot)
- [ ] https://github.com/maleadt/julia/blob/master/TODO.md
  - Refuse GC allocations
  - Refuse calling unknown functions
  - Allow codegen of ordinary arrays (`auto_unbox`?)
  - Allow multiple active codegen contexts. We could possibly generalize this,
    i.e. have 1 single context for host codegen, multiple ones (stack?) for PTX
    code, and select one in `emit_function` based on the active/parent target.
  - use `@target` to refuse cross-target calls
  - Implicit `return nothing`, and check for non-void kernels
  - Remove the old codegen exceptions for Array values (in the past we lowered
    normal Arrays to raw pointers, now we use an immutable `CuDeviceArray` for
    that purpose).
## Credits

Most of the work has been done by @maleadt and `Pieter Verstraete` see https://github.com/maleadt/julia/compare/master for a proper history. My part in the work was mostly squashing and rebasing the code onto upstream Julia, since @maleadt announced that he won't have time to work on this (see [post on julia-dev](https://groups.google.com/forum/#!topic/julia-dev/doQUqQabC8A)). I plan to continue working on this to get SPIR/SPIR-V support. Thanks to @SimonDanisch for helping with testing and motivating me. 
## How to try this

If you don't have a CUDA capable card you will need to use `gpuocelot` like described in http://blog.maleadt.net/2015/01/15/julia-cuda. Substitute CUDA.jl with my fork at https://github.com/vchuravy/CUDA.jl/tree/vc/fixes
